---
title: "About genofeatutil"
author: "Yen-Chung Chen"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette: 
    toc: true
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(genofeatutil)
```

## Introduction to `genofeatutil`

This is a collection of functions for working with Drosophila genomic datasets. Currently, it is mainly used to fetch gene names from objects created by `Seurat` and converting gene names, aliases, and FlyBase gene numbers to designated version of symbols of interest to sychronize the use of gene symbols in different datasets.


## General mode of action

### Prepare for every later conversion: `make_database()`

Gene name / number conversion functions rely on FlyBase. `make_database()` fetches the conversion table of aliases to up-to-date gene names and of deprecated FlyBase gene numbers from FlyBase, and creates a list object that can be used to efficently convert gene symbols or FlyBase numbers. Additionally, `make_database()` also requires a GTF file to specify the standard version of gene names to use, with which every conversion of gene names will end up (In constrast, conversion to FlyBase gene number gives gene numbers to the FlyBase database version specified in `make_database()`, which by default is the current version.) Available arguments including:
  1. `species`: Currently only `"dmel"` is supported, so it's not really on option.
  2. `version`: The version of FlyBase to use (e.g., `"FB2019_01"`).

```{r make-database, cache=TRUE}
# Fetch the current version of gene name and number from FlyBase
# Use the short BDGP6 GTF file in extdata folder as 
# an example target for later conversion
flybasedb <- make_database(species = "dmel", 
                           gtf.path = system.file("extdata", "short_bdgp6.gtf", 
                                                  package = "genofeatutil"))

str(flybasedb)
```

## Conversion

### To standardize gene names
Using the database generated by `make_database()`, `convert_to_genename()` converts gene names (including aliases) and FlyBase gene number to the gene name provided with the GTF file. Two arguments provide additional control to its behavior:
  1. `normalize`: If `TRUE`, the output will be prefixed with `gn_`, and punctuations not allowed (like parentheses or dash) in variable names will be replaced by ".". This behavior is nice when your expression matrix is a data frame of which the column names are gene names, because _column names are automatically converted to legit variable names_. To apply the same strategy of modifying gene names, see `?normalize_genename` and `?denormalize_genename` for more information.
  2. `remove.dup`: If `TRUE`, duplicated items will be removed from the output. This is useful when you are integrating many genes of interest without knowing if there are duplicated names under dinstinct aliases. If there are genes removed, a warning message will remind you.

```{r convert-gn-vanilla}
# No normalization and keep duplicates
convert_to_genename(c("Cyp307a2", "FBgn0086917", "spo2"), 
                    db = flybasedb, 
                    normalize = FALSE, 
                    remove.dup = FALSE)
```
```{r convert-gn-uniq}
# Remove duplicates
convert_to_genename(c("Cyp307a2", "FBgn0086917", "spo2"), 
                    db = flybasedb, 
                    normalize = FALSE, 
                    remove.dup = TRUE)
```
```{r convert-gn-standard}
# Remove duplicates and normalize the names
convert_to_genename(c("Cyp307a2", "FBgn0086917", "spo2"), 
                    db = flybasedb, 
                    normalize = TRUE, 
                    remove.dup = TRUE)
```

### From gene names to FlyBase gene numbers
`convert_gene_to_fbgn()` converts gene names and aliases to current version of FlyBase gene number if the database is fetched in default setting and the `version` argument is not customized. If the alias provided matches multiple FlyBase gene numbers, all matching numbers will be returned with a warning reminding you some gene names are not mapped 1-vs-1.

```{r to-fbgn}
convert_gene_to_fbgn(c("tj", "ap", "VGlut"), db = flybasedb)
```
```{r to-fbgn-exception}
convert_gene_to_fbgn(c("Cha", "spo2", "side"), db = flybasedb)
```

### Update outdated FlyBase gene names with `update_fbgn()`
`update_fbgn()` converts older FlyBase IDs to current counterpart. This is a partial equivalent of [Tools > Query by symbols/IDs > Upload/Convert IDs](http://flybase.org/convert/id). If there's no matching information for update, it would be assumed that the provided FlyBase gene number is uptodate.

```{r update-fbgn}
# dachs used to be FBgn0032045, but now it is FBgn0262029.
update_fbgn("FBgn0032045", db = flybasedb)
```