#' Scoring Predictors with a Selected Set of Criteria
#'
#' \code{score_predictors()} scores the predictor-target pair according to
#' the percentage of increment of MSE, the dominance of the predictive power (
#' represented by a gap in MSE among the predictors), correlation of expression
#' levels, and whether position-weighted matrix-inferred motif score of the
#' predictor is high near the target.
#'
#' To score the predictors, this function takes motif database from
#' \link[SCENIC] (motif analysis), an expression matrix (correlation analysis),
#' and a gene name conversion database from \code{\link{make_database}()}.
#'
#' @param x a data frame containing the predictors, targets, and
#' percentage increment of MSE (or other measures of predictor power)
#' @param gap.ratio a number indicating the threshold of "having a gap in MSE";
#' any %IncMSE beyond gap.ratio * the whole range of %IncMSE will be seen as a
#' gap and get +1 in score
#' @param mse.threshold a number indicating the threshold of individual
#' %IncMSE; any %IncMSE beyond this threshold gets +1 in score
#' @param cor.threshold a number indicating the threshold of individual
#' correlation of expression of the predictor-target pair; any correlation
#' beyond this thresold gets +1 in score
#' @param weight a numeric vector indicating the weights to sum up different
#' scores; the length is 7 with a default of c(1, 1, 1, 1, 1, 1, 1, 1). The
#' order of seven scores is "Raw%IncMSE", "gap", "MSE_threshold",
#' "cor_threshold", "Pearson_cor", "motif_score_sum", "motif_score_presence".
#' @param expMat a data frame containing an expression matrix or a character
#' string containing a path to a csv / rds / rdata file containing an
#' expression matrix
#' @param expMat.type a character string indicating the expMat file type if
#' \code{expMat} is a path
#' @param use.row a logical value indicating whether the gene names are in the
#' row names or column names of the expression matrix loaded
#' @param score.path a character string indicating the path to a SCENIC motif
#' score dataset
#' @param motif.path a character string indicating the path to a SCENIC motif
#' list
#' @param top.motif.number a numeric value indicating the number of top motifs
#' in terms of score to keep; the default is NULL and keeping all
#' @param motif.threshold a numeric value indicting the threshold score for
#' motifs; only motifs with scores higher than this value will be kept; the
#' default value is NULL and keeping all
#' @param db a list generated by \code{\link{make_database}()}
#' @param motif.list a list generated by \code{\link{get_motif_info()}}; if
#' this argument is not NULL, it has higher priority than \code{score.path} and
#' \code{motif.path}
#'
#' @return a data frame containing the scores of each target-predictor pair
#' @export
#'
#' @examples
#' # Loading dummydb
#' dummypath <- system.file("extdata", "dummy.gtf", package = "genofeatutil")
#' testdb <- make_database(species = "test", gtf.path = dummypath)
#'
#' # Generate dummy result for demo
#' dummyresult <- data.frame("predictor" = c("tf_1", "tf_2", "tf_3"),
#'                           "target" = rep("alias1.2", 3),
#'                           "Raw_%IncMSE" = c(0.3, 0.1, 1e-3),
#'                           row.names = NULL, stringsAsFactors = FALSE)
#'
#' dummyexpMat <- data.frame("sample1" = c(3, 1, 1, 3),
#'                           "sample2" = c(1, 1, 3, 2),
#'                           "sample3" = c(1, 3, 1, 1),
#'                           row.names = c("tf_1", "tf_2", "tf_3", "alias1.2"),
#'                           stringsAsFactors = FALSE)
#' # Example score data frame
#' score_result <- score_predictors(x = dummyresult,
#'                                  expMat = dummyexpMat,
#'                                  use.row = TRUE,
#'                                  score.path = system.file(
#'                                    "extdata",
#'                                    "scoremat.feather",
#'                                    package = "genofeatutil"),
#'                                  motif.path = system.file(
#'                                    "extdata",
#'                                    "dummy_motif.tbl",
#'                                    package = "genofeatutil"),
#'                                  db = testdb)
score_predictors <- function(x,
                             gap.ratio = 0.3,
                             mse.threshold = 0.1,
                             cor.threshold = 0.3,
                             weight = c(1, 1, 1, 1, 1, 1, 1),
                             expMat = NULL,
                             expMat.type = "csv",
                             use.row = TRUE,
                             motif.list = NULL,
                             score.path = NULL,
                             motif.path = NULL,
                             top.motif.number = NULL,
                             motif.threshold = NULL,
                             db = NULL) {
  # Type check
  if (class(x) != "data.frame") {
    stop("score_predictors() only takes a data.frame.\n")
  }

  # In case the input is not arranaged
  x <- x[order(x$Raw_.IncMSE, decreasing = TRUE), ]

  # Normalize gene names to be consistent
  x$predictor <- normalize_genename(x$predictor)
  x$target <- normalize_genename(x$target)

  # Gap score calculation
  full_range <- x$Raw_.IncMSE[1] - x$Raw_.IncMSE[nrow(x)]
  diff_mse <- diff(x$Raw_.IncMSE)
  ## Make everything before the gap TRUE
  score_gap <- diff_mse > gap.ratio * full_range
  is_gap <- match(TRUE, score_gap)
  if (!is.na(is_gap)) {
    score_gap[c(1:is_gap)] <- TRUE
  }
  gap_score <- as.numeric(c(score_gap, FALSE))

  # Check threshold of %IncMSE
  mse_score <- as.numeric(x$Raw_.IncMSE > mse.threshold)

  # Check if the predictor co-expresses with it's target
  if (!is.null(expMat)) {
    if (class(expMat) == "character") {
      avgexp <- versaread(expMat, type = expMat.type)
    }
    if (class(expMat) %in% c("data.frame", "matrix")) {
      avgexp <- expMat
    }
    if (!class(avgexp) %in% c("data.frame", "matrix")) {
      stop(paste("Please check if the path or data frame / matrix provided",
                 "is valid.\n"))
    }
    # Normalize gene names for expMat
    if (use.row) {
      row.names(avgexp) <- normalize_genename(row.names(avgexp))
    } else {
      colnames(avgexp) <- normalize_genename(colnames(avgexp))
    }


    cor_abs <- vector(length = nrow(x))
    for (rown in seq(nrow(x))) {
      cor_abs[rown] <- calculate_gene_cor(avgexp, gene1 = x$predictor[rown],
                                          gene2 = x$target[rown],
                                          use.row = use.row)
    }

    cor_score <- as.numeric(abs(cor_abs) > cor.threshold)
  }


  # Check if the target has a predictor's motif around it
  if (!is.null(score.path) & !is.null(motif.path)) {
    if (is.null(db)) {
      stop("Please make a database first with make_database().\n")
    }
    genes.use <- unique(x$target)

    motif_db <- get_motif_info(score_path = score.path,
                               motif_path = motif.path,
                               threshold = motif.threshold,
                               genes.use = genes.use,
                               number = top.motif.number,
                               db = db)

    motif_sum_score <- vector(length = nrow(x))
    motif_presence_score <- vector(length = nrow(x))

    for (rown in seq(nrow(x))) {
      this_motif <- motif_db[[x$target[rown]]]
      motif_presence_score[rown] <- as.numeric(
        x$predictor[rown] %in% this_motif)
      motif_sum_score[rown] <- sum(this_motif == x$predictor[rown])
    }
  }

  if (!is.null(motif.list)) {
    # Type check
    if (!is.list(motif.list)) {
      stop("motif.list should be a list generated by get_motif_info().\n")
    }

    motif_db <- motif.list

    motif_sum_score <- vector(length = nrow(x))
    motif_presence_score <- vector(length = nrow(x))

    for (rown in seq(nrow(x))) {
      this_motif <- motif_db[[x$target[rown]]]
      motif_presence_score[rown] <- as.numeric(
        x$predictor[rown] %in% this_motif)
      motif_sum_score[rown] <- sum(this_motif == x$predictor[rown])
    }
  }

  # Reverse modification for gene names
  x$predictor <- denormalize_genename(x$predictor)
  x$target <- denormalize_genename(x$target)



  # Generating a result table
  result_df <- data.frame("predictor" = x$predictor,
                          "target" = x$target,
                          "Raw_.IncMSE" = x$Raw_.IncMSE,
                          "gap" = gap_score,
                          "MSE_threshold" = mse_score,
                          "cor_threshold" = cor_score,
                          "Pearson_cor" = cor_abs,
                          "motif_score_sum" = motif_sum_score,
                          "motif_score_presence" = motif_presence_score,
                          row.names = NULL, stringsAsFactors = F
  )

  result_df$weighted_sum <- rowSums(sweep(x = result_df[ , 3:9], MARGIN = 2,
                                          FUN = `*`, STATS = weight))
  return(result_df)
}

integrate_score <- function(..., column) {

}

plot_score <- function() {

}
